### ISSUE TITLE
Distribute Game fragility; breakage due to gfx_sdl2 becoming default

### ISSUE THREAD

**Issue by rversteegen** [ohrrpgce/ohrrpgce#1185]
The Distribute Game options are very fragile, and now breaking. (Needs fixing before hróðvitnir). They download builds of either the latest stable release or latest nightly, and make assumptions about what files are in them and what builds they are. This means that when we change anything, it breaks all previous versions of the engine! This comes to a point now that gfx_sdl2 has become the default on Windows and Linux.

Wrong version downloaded
------------------------

The first problem is that they download a different version from what the user has, without even a warning. We don't have perfect back-compatibility, so we should at least warn. We could upload a .txt file with the name and date of the latest release.

But the "player"/"minimal" (warning: inconsistent naming, on Windows "minimal" includes Custom) downloads used by the distribmenu are archived at http://hamsterrepublic.com/ohrrpgce/archive/ anyway, so we *could* download the correct stable release instead of the latest one. Or easily give people the choice.

For nightlies I think it's acceptable to just download the latest nightly, or they could get the next stable release after their current nightly, if we had a list of releases in a .txt (or check the archive directory listing).

Build assumptions
-----------------

For Windows, distribmenu assumes which dlls are in the .zip and need to be included.
`get_windows_gameplayer` is hardcoded to extract SDL2.dll and SDL2_mixer.dll from the zip; `find_required_dlls` also needs changing. `write_innosetup_script` just includes whichever .dlls were extracted from the zip. If we simply change them all to just include all the files in the .zip that seems good enough.

For Linux, distribmenu assumes to know what the dependencies for .deb files are. (Currently still wrongly set to sdl 1.2.)
I suggest we put a .txt/ini in the download that lists the current dependencies.
(In future if we include libsdl2.so and libsdl2_mixer.so and package them in tarballs (but not .debs) then things change again; can't just include all files.)

For Mac distribmenu doesn't assume what the build is, so switching to gfx_sdl2 won't break anything, but it does make assumptions about how to put an .rpg and .icns inside, which might change some day.

On all platforms, I suggest we put a .txt (or .ini) in the downloads containing a "package format version" number and other meta information, so old versions of Custom can display "You need to download a more recent engine version" if we change anything.

Currently, once hróðvitnir is released, distributing for Windows will break in all previous stable releases (when game.exe needs to be downloaded). The URL should be changed.
I fixed this for nightly builds already by uploading ohrrpgce-player-win-wip-sdl2.zip instead of ohrrpgce-player-win-wip.zip.

Stable releases download

ohrrpgce-player-win.zip
ohrrpgce-player-linux-bin-minimal.zip
ohrrpgce-player-linux-bin-minimal-x86_64.zip
ohrrpgce-mac-minimal-x86.tar.gz
ohrrpgce-mac-minimal-x86_64.tar.gz
from
http://hamsterrepublic.com/dl/

Nightly releases download

ohrrpgce-player-win-wip-sdl2.zip  (previously ohrrpgce-player-win-wip.zip)
ohrrpgce-player-linux-bin-minimal.zip
ohrrpgce-player-linux-bin-minimal-x86_64.zip
ohrrpgce-mac-minimal-x86.tar.gz
ohrrpgce-mac-minimal-x86_64.tar.gz
from
http://hamsterrepublic.com/ohrrpgce/nightly/

I suggest we give people the choice of gfx_sdl 32-bit linux builds or gfx_sdl2 64-bit linux builds, and put the gfx_sdl2 builds at ohrrpgce-player-linux-bin-minimal-x86_64-sdl2.zip, so old releases can continue to download gfx_sdl builds from the old URLs and produce correct .debs.
(Or rename both linux zips, to make the naming more consistent at the same time!)

New/Reorganised Distribute Options
----------------------

Because gfx_sdl2 is Win XP+, we should add an "Export .zip for Obsolete Windows" option in the Menu for Win 95+ which packages using a gfx_sdl build.

I want to reorganise the menu into two halves: main export options (WindowsXP+ zip/installer, 64-bit Mac, 64-bit Linux tarball/deb), and options for obsolete OSes (gfx_sdl builds for Windows, 32-bit Mac and 32-bit Linux).

**Comment by ohrrpgce-githubmail** [comment]
I wish I had thought out these issues sooner, but I do think that your
plans for mitigating this problems are good. I have a shortage of free time
in the next few weeks, so I can't start on these right away, but I will
help when I can.

I wonder how helpful it would be if I added a PHP file that would dump a
listing of downloads in a more parseable format, rather than trying to
parse the html of an Apache directory listing?

**Comment by bob-the-hamster** [comment]
I was trying to break this bug down into a minimal list of fixes to un-block the hróðvitnir release, and I got lost pretty quick.

We need to spit this bug into a minimum of 3 different bugs

1) Things that must change to prevent breakage in fufluns and gorgonzola when hróðvitnir releases (bonus points for preventing breakage in older versions too, but I would only consider the last 2 as blockers) -  #1204 

2) Things that must change to un-block the hróðvitnir stable release

3) Improvements that are nice-to-have going forward, but don't block the release

I think all 3 of those are likely to include a mixture of changes to the distribute game menu in custom, and changes to the layout of files on hamsterrepublic.com

EDIT: I guess the future improvements should just stay in this bug, but I'll still make two more bugs for 1) and 2) and link them in here

**Comment by bob-the-hamster** [comment]
I made a few fixes for this issue
* Warning that a newer version of the ohrrpgce than the one you are currently using might be downloaded
* Fix dll inclusion logic for windows distrib to use whatever was in the download, instead of hardcoding SDL2.dll and SDL2_mixer.dll

**Comment by rversteegen** [comment]
Those changes look adequate to unblock release, except that `write_debian_control_file` still is broken, which I apparently forgot to mention as a blocker. Maybe when you're packaging .debs the warning "The latest <foo> version of the OHRRPGCE will be used, even if that is newer than the version you are currently using" needs to be appended with "If the version differs, the resulting .deb file may be broken". Otherwise, we'd need to put a `buildinfo.ini` file in the download, and get the .deb requirements from that.

**Comment by bob-the-hamster** [comment]
All of the release-blocking problems from this bug have been fixed.

I'm going to close this bug, if we need further fixes/improvements to the Distribute Game feature, lets open separate smaller issues for each.
